homeassistant/components/ecovacs_mqtt/entity.py

```diff
@@ -0,0 +1,117 @@
+"""Ecovacs mqtt entity module."""
+from __future__ import annotations
+
+from collections.abc import Callable, Coroutine
+from dataclasses import dataclass
+from typing import Any, Generic, TypeVar
+
+from deebot_client.capabilities import Capabilities
+from deebot_client.device import Device
+from deebot_client.events import AvailabilityEvent
+from deebot_client.events.base import Event
+
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity import Entity, EntityDescription
+
+from .const import DOMAIN
+
+_EntityDescriptionT = TypeVar("_EntityDescriptionT", bound=EntityDescription)
+CapabilityT = TypeVar("CapabilityT")
+EventT = TypeVar("EventT", bound=Event)
+
+
+@dataclass(kw_only=True, frozen=True)
+class EcovacsEntityDescription(
+    EntityDescription,
+    Generic[CapabilityT],
+):
+    """Ecovacs entity description."""
+
+    capability_fn: Callable[[Capabilities], CapabilityT | None]
+
+
+class EcovacsEntity(Entity, Generic[CapabilityT, _EntityDescriptionT]):
+    """Ecovacs entity."""
+
+    entity_description: _EntityDescriptionT
+
+    _attr_should_poll = False
+    _attr_has_entity_name = True
+    _always_available: bool = False
+    _subscribed_events: set[type[Event]] = set()
+
+    def __init__(
+        self,
+        device: Device,
+        capability: CapabilityT,
+        entity_description: _EntityDescriptionT | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize entity."""
+        super().__init__(**kwargs)
+        if entity_description:
+            self.entity_description = entity_description
+        elif not hasattr(self, "entity_description"):
+            raise ValueError(
+                '"entity_description" must be either set as class variable or passed on init!'
+            )
+
+        self._device = device
+        self._capability = capability
+
+        self._attr_unique_id = self._device.device_info.did
+
+        if self.entity_description.key:
+            self._attr_unique_id += f"_{self.entity_description.key}"
```

Comment:
```suggestion
            self._attr_unique_id += f"_{entity_description.key}"
```

Reply:
Must use `self.entity_description` here as entity description can be set also as class variable